internal/pkg/action/dispatcher.go

```diff
@@ -96,31 +97,63 @@ func (d *Dispatcher) process(ctx context.Context, hits []es.HitT) {
 			log.Error().Err(err).Msg("Failed to unmarshal action document")
 			break
 		}
-		for _, agentId := range action.Agents {
-			arr := agentActions[agentId]
+		numAgents := len(action.Agents)
+		for i, agentID := range action.Agents {
+			arr := agentActions[agentID]
 			actionNoAgents := action
+			actionNoAgents.StartTime = offsetStartTime(action.StartTime, action.Expiration, action.MinimumExecutionDuration, i, numAgents)
 			actionNoAgents.Agents = nil
 			arr = append(arr, actionNoAgents)
-			agentActions[agentId] = arr
+			agentActions[agentID] = arr
 		}
 	}
 
-	for agentId, actions := range agentActions {
-		d.dispatch(ctx, agentId, actions)
+	for agentID, actions := range agentActions {
+		d.dispatch(ctx, agentID, actions)
 	}
 }
 
-func (d *Dispatcher) getSub(agentId string) (Sub, bool) {
+// offsetStartTime will return a new start time between start:end-dur based on index i and the total number of agents
+// An empty string will be returned if start or exp are empty or if there is an error parsing inputs
+// As we expect i < total  the latest return time will always be < exp-dur
+func offsetStartTime(start, exp, dur string, i, total int) string {
+	if start == "" || exp == "" {
+		return ""
+	}
+	startTS, err := time.Parse(time.RFC3339, start) // TODO what format does a date-time string use?
```

Comment:
Shouldn't we move the validation to the original `unpack` of the action instead? If there is an `unserialization error` this means that the contract between fleet and fleet-server is broken. I don't think we should even try to process that event.

Reply:
To summarize a quick conversation. We should catch this on the json unpacking, however the `model.Action` is generated by `elastic/go-json-schema-generate` which does not support mapping a (json) `string/date-time` to go's `time.Time`. The upstream generator we forked our code from has this as an open issue https://github.com/a-h/generate/issues/69 so if we get around to it we should contribute it back